File: C++/src/Sum.cpp
//
// Created by kde on 4/13/16.
//

#include "Sum.h"
#include <iostream>

Sum::Sum() {

}

Sum::Sum(IOperation * firstArg_, IOperation * secondArg_) {
  Sum::setArguments(firstArg_, secondArg_);
}

double Sum::evaluate() {
  return firstArg->evaluate() + secondArg->evaluate();
}

void Sum::setArguments(IOperation * firstArg_, IOperation * secondArg_) {
  Sum::firstArg = firstArg_;
  Sum::secondArg = secondArg_;
}

void Sum::toString() {
  std::cout << "Sum two values" << ", fa: " << firstArg << "\n";
}

IOperation * Sum::firstArgument() {
  return firstArg;
}
IOperation * Sum::secondArgument() {
  return secondArg;
}

File: C#/calculator/calculator/Sum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace calculator {
    class Sum : BinaryOperation {
        private readonly IOperation _firstOperation; // { get; }
        private readonly IOperation _secondOperation; // { get; }

        public Sum(IOperation firstOperation, IOperation secondOperation) {
            this._firstOperation = firstOperation;
            this._secondOperation = secondOperation;
        }

        public override double Evaluate() {
            return _firstOperation.Evaluate() + _secondOperation.Evaluate();
        }

        public override string ToString() {
            String firstString = this._firstOperation.ToString();
            String secondString = this._secondOperation.ToString();

            if (this._firstOperation.GetDepth() > 0) {
                firstString = "(" + firstString + ")";
            }
            if (this._secondOperation.GetDepth() > 0) {
                secondString = "(" + secondString + ")";
            }

            return firstString + " + " + secondString;
        }

        protected override IOperation FirstArgument() {
            return this._firstOperation;
        }

        protected override IOperation SecondArgument() {
            return this._secondOperation;
        }
    }
}
